test/test_autodiff.cpp

```diff
@@ -0,0 +1,1093 @@
+            else if (i==0 && j==1)
+                BOOST_REQUIRE(empty.derivative(i,j) == 1.0);
+            else
+                BOOST_REQUIRE(empty.derivative(i,j) == 0.0);
+    empty = cx; // set a constant
+    for (int i=0 ; i<=m ; ++i)
+        for (int j=0 ; j<=n ; ++j)
+            if (i==0 && j==0)
+                BOOST_REQUIRE(empty.derivative(i,j) == cx);
+            else
+                BOOST_REQUIRE(empty.derivative(i,j) == 0.0);
+}
+
+BOOST_AUTO_TEST_CASE(addition_assignment)
+{
+    constexpr int m = 3;
+    constexpr int n = 4;
+    constexpr double cx = 10.0;
+    auto sum = boost::math::differentiation::autodiff::variable<double,m,n>(); // zero-initialized
+    // Single variable
+    const auto x = boost::math::differentiation::autodiff::variable<double,m>(cx);
+    sum += x;
+    for (int i=0 ; i<=m ; ++i)
+        for (int j=0 ; j<=n ; ++j)
+            if (i==0 && j==0)
+                BOOST_REQUIRE(sum.derivative(i,j) == cx);
+            else if (i==1 && j==0)
+                BOOST_REQUIRE(sum.derivative(i,j) == 1.0);
+            else
+                BOOST_REQUIRE(sum.derivative(i,j) == 0.0);
+    // Arithmetic constant
+    constexpr double cy = 11.0;
+    sum = 0;
+    sum += cy;
+    for (int i=0 ; i<=m ; ++i)
+        for (int j=0 ; j<=n ; ++j)
+            if (i==0 && j==0)
+                BOOST_REQUIRE(sum.derivative(i,j) == cy);
+            else
+                BOOST_REQUIRE(sum.derivative(i,j) == 0.0);
+}
+
+BOOST_AUTO_TEST_CASE(subtraction_assignment)
+{
+    constexpr int m = 3;
+    constexpr int n = 4;
+    constexpr double cx = 10.0;
+    auto sum = boost::math::differentiation::autodiff::variable<double,m,n>(); // zero-initialized
+    // Single variable
+    const auto x = boost::math::differentiation::autodiff::variable<double,m>(cx);
+    sum -= x;
+    for (int i=0 ; i<=m ; ++i)
+        for (int j=0 ; j<=n ; ++j)
+            if (i==0 && j==0)
+                BOOST_REQUIRE(sum.derivative(i,j) == -cx);
+            else if (i==1 && j==0)
+                BOOST_REQUIRE(sum.derivative(i,j) == -1.0);
+            else
+                BOOST_REQUIRE(sum.derivative(i,j) == 0.0);
+    // Arithmetic constant
+    constexpr double cy = 11.0;
+    sum = 0;
+    sum -= cy;
+    for (int i=0 ; i<=m ; ++i)
+        for (int j=0 ; j<=n ; ++j)
+            if (i==0 && j==0)
+                BOOST_REQUIRE(sum.derivative(i,j) == -cy);
+            else
+                BOOST_REQUIRE(sum.derivative(i,j) == 0.0);
+}
+
+BOOST_AUTO_TEST_CASE(multiplication_assignment)
+{
+    constexpr int m = 3;
+    constexpr int n = 4;
+    constexpr double cx = 10.0;
+    auto product = boost::math::differentiation::autodiff::variable<double,m,n>{1}; // unit constant
+    // Single variable
+    auto x = boost::math::differentiation::autodiff::variable<double,m>(cx);
+    product *= x;
+    for (int i=0 ; i<=m ; ++i)
+        for (int j=0 ; j<=n ; ++j)
+            if (i==0 && j==0)
```

Comment:
Again, I personally prefer explicit bracing. Don't know if this is bikeshedding at the point or not.

Reply:
Normally, yes :)  But for lots of nested loops I would agree with Nick.